tests/integration_tests/tests/agentless_tests/test_blueprint_upload.py

```diff
@@ -0,0 +1,242 @@
+        self.client.blueprints.publish_archive(
+            archive_url,
+            blueprint_id)
+        wait_for_blueprint_upload_to_end(blueprint_id, self.client)
+        blueprint = self.client.blueprints.get(blueprint_id)
+
+        # This is caught at `upload_archive_to_file_server`, before extracting
+        self.assertEqual(blueprint['state'],
+                         BlueprintUploadState.FAILED_UPLOADING)
+        self.assertRegex(
+            blueprint['error'],
+            "Blueprint archive is of an unrecognized "
+            "format {}".format(archive_url))
+
+    def test_blueprint_upload_from_url_invalid_archive_structure(self):
+        blueprint_id = 'bp-url-bad-structure'
+        archive_url = 'https://cloudify-tests-files.s3-eu-west-1.amazonaws' \
+                      '.com/blueprints/not-a-valid-archive.zip'
+        self.client.blueprints.publish_archive(
+            archive_url,
+            blueprint_id)
+        wait_for_blueprint_upload_to_end(blueprint_id, self.client)
+        blueprint = self.client.blueprints.get(blueprint_id)
+
+        self.assertEqual(blueprint['state'],
+                         BlueprintUploadState.FAILED_EXTRACTING)
+        self.assertRegex(
+            blueprint['error'],
+            "Archive must contain exactly 1 directory")
+
+    def test_blueprint_upload_from_url_missing_yaml(self):
+        blueprint_id = 'bp-url-missing-yaml'
+        blueprint_filename = 'fancy_name.yaml'
+        archive_url = 'https://cloudify-tests-files.s3-eu-west-1.amazonaws' \
+                      '.com/blueprints/the-not-blueprint-master.zip'
+        self.client.blueprints.publish_archive(
+            archive_url,
+            blueprint_id,
+            blueprint_filename=blueprint_filename
+        )
+        wait_for_blueprint_upload_to_end(blueprint_id, self.client)
+        blueprint = self.client.blueprints.get(blueprint_id)
+
+        self.assertEqual(blueprint['state'],
+                         BlueprintUploadState.FAILED_EXTRACTING)
+        self.assertRegex(blueprint['error'],
+                         '{0} does not exist in the application '
+                         'directory'.format(blueprint_filename))
+
+    def test_blueprint_reupload_after_fail(self):
+        blueprint_id = 're-bp'
+        self.client.blueprints.upload(resource('dsl/basic.yaml'),
+                                      entity_id=blueprint_id)
+        wait_for_blueprint_upload_to_end(blueprint_id, self.client)
+        blueprint = self.client.blueprints.get(blueprint_id)
+
+        # this should fail due to cloudmock plugin not uploaded
+        self.assertEqual(blueprint['state'], BlueprintUploadState.INVALID)
+        self.assertEqual(blueprint.plan, None)
+        self.assertRegex(blueprint['error'],
+                         'Plugin cloudmock .* not found')
+        original_creation_time = blueprint['created_at']
+
+        self.client.blueprints.upload(resource('dsl/empty_blueprint.yaml'),
```

Comment:
I think this step should upload the same blueprint to simulate real user behavior. This means the cloudmock plugin needs to be uploaded beforehand.

Reply:
uploading a plugin for which we should build a wagon mid-test introduces unnecessary complexity for what the test checks. Fixing the plugin file itself after discovering it has a bad import is realistic